File: Tps/tp01/ex17/PaliRec.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <locale.h>

bool ehFim(char str[]){
    //Função que verifica se a entrada é o FIM
    //Inicialmente apenas compara se a entrada tem o tamanho certo para ser FIM depois verifica string
    bool ehFim = false;
    if(strlen(str) == 3){
        if(str[0] == 'F' && str[1] == 'I' && str[2] == 'M'){
            ehFim = true;
        }
    }
    return ehFim;
}

bool verificaPalindromo(char entrada[], int fim,int inicio){
        // A variavel ja começa verdadeira, pois no primeiro falso acaba
		bool iguais = true;
		if(fim > inicio){
            // Verifica cada letra em busca de diferença, caso ocorra iguais vira falso e encerra a função
            if(entrada[inicio] != entrada[fim]){
                iguais = false;
            }
            else{
                verificaPalindromo(entrada, fim-1, inicio+1);
            }
        }
        return iguais;
}

int main(void){
    // Trocando a linguagem para ler as entradas
	setlocale(LC_ALL, "pt_BR.UTF-8");
    // Declarando a variavel que ira armazenar a string
	char str[10000];
	fgets(str, sizeof(str), stdin);
    // Removendo o \n do final da string para não dar erro na comparação
	str[strcspn(str, "\n")] = '\0';
	int tamanho;
	while(!ehFim(str))
	{
        // Tamanho recebe strlen - 1 para ser usado como indice
		tamanho = strlen(str) - 1;
		if(verificaPalindromo(str, tamanho, 0)){
			printf("SIM\n");
		}
		else{
			printf("NAO\n");
		}
        // Repetindo o processo de leitura da string
		fgets(str, sizeof(str), stdin);
		str[strcspn(str, "\n")] = '\0';
	}
}

File: Tps/tp01/ex16/PaliRec.java
import java.util.*;
public class PaliRec{
    public static boolean ehFim(String entrada) {
        //Função que verifica se a entrada é o FIM
        //Inicialmente apenas compara se a entrada tem o tamanho certo para ser FIM depois verificaa string
        boolean fim = false;
        if (entrada.length() == 3) {
            if ((entrada.charAt(0) == 'F') && (entrada.charAt(1) == 'I') && (entrada.charAt(2) == 'M')) {
                fim = true;
            }
        }
        return fim;
    }

    public static boolean chamaVerifica(String entrada, int tamanho){
        // Função que chama a outra função para evitar erros
        return verificaPali(entrada, 0, tamanho);
    }

    public static boolean verificaPali(String entrada, int inicio, int fim){
        // Incia a variavel considerando que é verdadeira
        boolean ehPalindromo = true;
        if(fim > inicio){
            // Caso houver diferença em alguma posição, troca o bool para falso e encerra a função
            if(entrada.charAt(fim) != entrada.charAt(inicio)){
                ehPalindromo = false;
            }
            else{
                // A função so se repete quando as letras forem iguais
                verificaPali(entrada, inicio+1, fim-1);
            }
        }
        return ehPalindromo;
    }

    public static void main(String args[]) {
        // Trocando o charset atual para ler corretamente as entradas
        Scanner ler = new Scanner(System.in, "UTF-8");
        String palavra = ler.nextLine();
        while (!ehFim(palavra)) {
            //Crio uma variavel que recebe o tamanho - 1, para ja ser utilizada como índice
            int tamanho = palavra.length() - 1;
            //Imprime a resposta da linha e ja recebe um nova string
            if (chamaVerifica(palavra, tamanho)) {
                System.out.println("SIM");
            } else {
                System.out.println("NAO");
            }
            palavra = ler.nextLine();
        }
    }
}